includes/models/model.llms.add-on.php

```diff
@@ -452,35 +452,58 @@ private function lookup_add_on( $lookup_key, $lookup_val ) {
 	}
 
 	/**
-	 * Uninstall an add-on and permanently deletes its files
+	 * Verifies the add-on can be uninstalled, and performs the uninstall (permanently deleting its files)
 	 *
 	 * @since [version]
 	 *
-	 * @return string|WP_Error
+	 * @return string|WP_Error Success message or an error object.
 	 */
 	public function uninstall() {
 
-		$ret = sprintf( __( '%s was successfully uninstalled.', 'lifterlms' ), $this->get( 'title' ) );
+		$title = $this->get( 'title' );
 
 		if ( ! $this->is_installed() ) {
-			$ret = new WP_Error( 'not-installed', sprintf( __( '%s is not installed.', 'lifterlms' ), $this->get( 'title' ) ) );
-		} elseif ( $this->is_active() ) {
-			$ret = new WP_Error( 'uninstall-active', sprintf( __( '%s is active and cannot be uninstalled.', 'lifterlms' ), $this->get( 'title' ) ) );
+			// Translators: %s = Add-on title.
+			return new WP_Error( 'not-installed', sprintf( __( '%s is not installed.', 'lifterlms' ), $title ) );
+		}
+
+		if ( $this->is_active() ) {
+			// Translators: %s = Add-on title.
+			return new WP_Error( 'uninstall-active', sprintf( __( '%s is active and cannot be uninstalled.', 'lifterlms' ), $title ) );
+		}
+
+		return $this->uninstall_real();
+
+	}
+
+	/**
+	 * Actually performs the uninstall
+	 *
+	 * @since [version]
+	 *
+	 * @return string|WP_Error Success message or an error object.
+	 */
+	private function uninstall_real() {
+
+		$type = $this->get_type();
+		$file = $this->get( 'update_file' );
+		$del  = null;
+		if ( 'plugin' === $type ) {
+			uninstall_plugin( $file );
+			$del = delete_plugins( array( $file ) );
+		} elseif ( 'theme' === $type ) {
+			$del = delete_theme( $file );
 		} else {
+			// Translators: %s = add-on type.
+			return new WP_Error( 'uninstall-invalid-type', sprintf( __( 'Cannot uninstall "%s" type add-ons.', 'lifterlms' ), $type ) );
+		}
```

Comment:
Alright, I thought about that but I kinda trusted that you didn't take care of this 3rd case (invalid type) because it couldn't happen.
About the code itself, what do you think if we early bail when the type is not a valid one, to avoid if/elseif/else that are a bit ugly to me. I mean make it like:
```php
		$type = $this->get_type();
		if ( ! in_array( $type, array( 'theme', 'plugin' ), true ) ) {
			// Translators: %s = add-on type.
			return new WP_Error( 'uninstall-invalid-type', sprintf( __( 'Cannot uninstall "%s" type add-ons.', 'lifterlms' ), $type ) );
		}

		$file = $this->get( 'update_file' );

		if ( 'plugin' === $type ) {
			uninstall_plugin( $file );
			$del = delete_plugins( array( $file ) );
		} else {
			$del = delete_theme( $file );
		} 

```